### DISCUSSION TITLE
Documentation about template-strings

### DISCUSSION THREAD

**Discussion by OskarKlintrot** [spectreconsole/spectre.console#270] [General]
Is there any documentation about the template for `CommandArgumentAttribute` etc? The hint just says `The argument template.` but I have no clue what that means. Does the `[]` or `<>` actually mean/do anything? What else can we use? I guess you can figure it out by reading the source code for [TemplateTokenizer](https://github.com/spectresystems/spectre.console/blob/3a42c0a119af659c842922ca8759b6a63862a876/src/Spectre.Console/Cli/Internal/Configuration/TemplateTokenizer.cs) but a more convenient way would be appreciated :)

**Comment by patriksvensson** [comment]
@OskarKlintrot Excellent question. I'm a bit behind on documenting Spectre.Console.Cli, but to summarize:

`[FOO]` indicates an optional value
`<FOO>` indicates a required value.

```csharp
public class ConsoleSettings : CommandSettings
{
    [CommandArgument(0, "<MANDATORY>")]
    [Description("Mandatory argument")]
    public string Mandatory { get; set; } = string.Empty;

    [CommandArgument(1, "[OPTIONAL]")]
    [Description("Optional argument")]
    public string? Optional { get; set; }

    [CommandOption("--command-option-flag")]
    [Description("Command option flag.")]
    public bool CommandOptionFlag { get; set; }

    [CommandOption("--command-option-value|-c1|--command-req <REQUIRED_VALUE>")]
    [Description("Command option value (required).")]
    public string? CommandOptionValue { get; set; }

    [CommandOption("--command-option-value-opt|-c2|--command-opt [OPTIONAL_VALUE]")]
    [Description("Command option value (optional).")]
    public FlagValue<string> CommandOptionValueOptional { get; set; }
}
```

**Reply by Goffen** [reply]
@patriksvensson  I don't get it! How do you the CLI to validate?

When I change in the "Demo" project the "RunCommand" like this: 

```
public override int Execute(CommandContext context, Settings settings)
{
    if (!settings.Validate().Successful) { throw new Exception("NOT OK");}
```
And run it without anything except `run`. I expect it to do some kind of validation?

**Reply by Goffen** [reply]
Looking in your example code. Why nullable on CommandOptionValue, is it not required (because of <>)?

And why three things between pipes "short | long | ..whatthis?"

**Reply by OskarKlintrot** [reply]
@Goffen The Demo project validates successfully, all options are optional. 

`string?` is nullable since it's optional (as far as I can tell only `CommandArgument` can be required). `<>` only makes the value required _if_ you use the flag.

`|` is used for giving the flag multiple names. So to set `CommandOptionValueOptional` you can use any of the flags `--command-option-value-opt`, `-c2` or `--command-opt`. I usually set one descriptive flag and one short. So `--environment` and `-e`: `-e|--environment`.

I have no clue though how `FlagValue` works, can't find anything about it. Hopefully @patriksvensson can add some clarity there.

**Reply by Goffen** [reply]
Aha only `CommandArgument` can be required?! That might explain it. Then it's wierd that you can use angle brackets `<>` in `CommandOption`

**Reply by patriksvensson** [reply]
@Goffen You can have a flag that can have an optional value using `FlagValue<T>` like shown here: https://github.com/spectreconsole/spectre.console/blob/main/examples/Cli/Demo/Commands/Serve/ServeCommand.cs#L17-L19

```csharp
[CommandOption("-o|--open-browser [BROWSER]")]
[Description("Open a web browser when the server starts. You can also specify which browser to use. " + 
             "If none is specified, the default one will be used.")]
public FlagValue<string> OpenBrowser { get; set; }
```

**Reply by Goffen** [reply]
@patriksvensson you mean @OskarKlintrot? I get how FlagValue is used, more or less like `?`. But how do I get Spectre to validate required `CommandOptions`?

**Reply by Goffen** [reply]
`--command-req` the third parameter in the '|' thingy, is that how you get it to be required?

**Reply by patriksvensson** [reply]
@Goffen Options are, well options. They can't be required per se. 
The only thing we can validate is that if you pass an option with a required value, and don't provide one.

```
[CommandOption("--foo <THE_VALUE>")]
public string Foo { get; set; }
```

If I run the application with only `--foo`, that will fail validation.
Validation of settings is done automatically by the framework, so you don't need to call the Validate method yourself.

## Custom validation - Overrides

You can provide your own validation code by overriding the `Validate` method in the command.

```csharp
public override ValidationResult Validate(
    [NotNull] CommandContext context,
    [NotNull] Settings settings)
{
    //  Custom validation
    if (settings.Foo % 2 != 0)
    {
        return ValidationResult.Error("Number is not even");
    }

    // Return success
    return ValidationResult.Success();
}
```

## Custom validation - Attributes

Or you can implement your own validation attribute and decorate the Option property

```csharp
[AttributeUsage(AttributeTargets.Property, AllowMultiple = true)]
public sealed class EvenNumberValidatorAttribute : ParameterValidationAttribute
{
    public EvenNumberValidatorAttribute(string errorMessage)
        : base(errorMessage)
    {
    }

    public override ValidationResult Validate(CommandParameterContext context)
    {
        if (context.Value is int integer)
        {
            if (integer % 2 == 0)
            {
                return ValidationResult.Success();
            }

            return ValidationResult.Error($"Number is not even ({context.Parameter.PropertyName}).");
        }

        throw new InvalidOperationException($"Parameter is not a number ({context.Parameter.PropertyName}).");
    }
}
```

And then in your settings class

```csharp
[CommandOption("--foo <THE_VALUE>")]
[EvenNumberValidator]
public int Foo { get; set; }
```

**Reply by Goffen** [reply]
Aaahhhh! Okey. That's kind of strange behaviour to have it only be required if you supply it ;) 

And with CommandArguments you cannot use the `-f|--foo` notation. 

For example the azure commandline client uses the -x|-xxx for required things. I find it very useful because parameters are named, instead if they only used CommandArgument you need to remember the ordering of everything. 

For example
`az group create --location westeurope --name myResourceGroup`

is way better than 

`az group create westeurope myResourceGroup`

**Reply by woutervanranst** [reply]
@Goffen this is _the_ reason I stayed with System.Commandline... you can have mandatory required arguments there
https://github.com/dotnet/command-line-api/blob/main/docs/Your-first-app-with-System-CommandLine.md

**Reply by VictorioBerra** [reply]
Just landed here from Google, same exact gripe @patriksvensson, naming arguments is incredibly important. Is there no way we can introduce a required flag on `[CommandOption(Required = true)]`? If Required is true, the option MUST be provided. Or, allow named arguments.

**Reply by FrankRay78** [reply]
`CommandLine` does indeed support required options, as per: https://learn.microsoft.com/en-us/dotnet/standard/commandline/get-started-tutorial

Here's an example from their documentation:

```csharp
        var searchTermsOption = new Option<string[]>(
            name: "--search-terms",
            description: "Strings to search for when deleting entries.")
            { IsRequired = true, AllowMultipleArgumentsPerToken = true };
```

I can see the logic behind explicitly named options on the command line, for readability.

@patriksvensson - Do you have thoughts about whether `spectre.console` should follow suit, or not?
